server/src/server.ts

```diff
@@ -9,306 +9,378 @@ import { BufWriter } from './protocol/BufWriter'
 			} catch (err) {
 				this.warn(err)
-				return this.kick('Error in data handler')
+				this.kick('Error in data handler')
+				return
 			}
 		})
-
 		socket.on('close', (hadError: boolean) => {
 			this.log('Closed.', { hadError })
 		})
-
 		socket.on('end', () => {
 			this.log('Ended')
 		})
-
 		socket.on('timeout', () => {
 			this.warn('Timeout')
 		})
-
 		socket.on('error', (err: Error) => {
 			this.warn('Error:', err)
 			this.kick('Socket error')
 		})
 	}
 
-	private async handlePacketReceived(pkt: ClientPacket) {
-		if (!this.uuid) {
-			// not authenticated yet
-			switch (pkt.type) {
-				case 'Handshake':
-					return await this.handleHandshakePacket(pkt)
-				case 'EncryptionResponse':
-					return await this.handleEncryptionResponsePacket(pkt)
-			}
-			throw new Error(`Packet ${pkt.type} from unauth'd client ${this.id}`)
-		} else {
-			return await this.handler.handleClientPacketReceived(this, pkt)
-		}
-	}
-
-	kick(internalReason: string) {
-		this.log(`Kicking:`, internalReason)
+	/**
+	 * Forcibly disconnects the client.
+	 *
+	 * @param internalReason The reason for the disconnect. This is NOT sent to the client.
+	 */
+	public kick(internalReason: string) {
+		this.log('Kicking:', internalReason)
 		this.socket.destroy()
 	}
 
-	async send(pkt: ServerPacket) {
+	/**
+	 * Sends a packet to the client.
+	 *
+	 * @param packet The packet to send.
+	 */
+	public async send(packet: ServerPacket) {
 		if (!this.cryptoPromise) {
-			this.debug('Not encrypted, dropping packet', pkt.type)
+			this.debug('Not encrypted, dropping packet', packet.type)
 			return
 		}
-		if (!this.uuid) {
-			this.debug('Not authenticated, dropping packet', pkt.type)
+		if (!this.mcPlayerUuid) {
+			this.debug('Not authenticated, dropping packet', packet.type)
 			return
 		}
-		if (pkt.type !== 'ChunkTile') this.debug(this.mcName + " -> " + pkt.type);
-		await this.sendInternal(pkt, true)
+		if (packet.type !== 'ChunkTile') {
+			this.debug(this.mcPlayerName + ' -> ' + packet.type)
+		}
+		await this.INTERNAL_send(packet, true)
 	}
 
-	private async sendInternal(pkt: ServerPacket, doCrypto = false) {
-		if (!this.socket.writable)
-			return this.debug('Socket closed, dropping', pkt.type)
-		if (doCrypto && !this.cryptoPromise)
-			throw new Error(`Can't encrypt: handshake not finished`)
-
-		const writer = new BufWriter() // TODO size hint
-		writer.writeUInt32(0) // set later, but reserve space in buffer
-		encodePacket(pkt, writer)
-		let buf = writer.getBuffer()
-		buf.writeUInt32BE(buf.length - 4, 0) // write into space reserved above
-
+	private async INTERNAL_send(packet: ServerPacket, doCrypto: boolean = false) {
+		if (!this.socket.writable) {
+			this.debug('Socket closed, dropping', packet.type)
+			return
+		}
+		if (doCrypto && !this.cryptoPromise) {
+			throw new Error("Can't encrypt: handshake not finished")
+		}
+		const packetWriter = new BufWriter() // TODO: size hint (Ori: presumably means reducing unnecessary expansions)
```

Comment:
your presumption is correct, if we knew a typical upper bound of packet sizes we could save some buffer allocations here